### DISCUSSION TITLE
Implementing a Debug Renderer

### DISCUSSION THREAD

**Discussion by cuppajoeman** [jrouwe/JoltPhysics#902] [General]
Hi I'm working with jolt to create my game, I need to get a jolt debug renderer going so I can actually see what jolt sees.

I'm using opengl3 and I have been looking at https://jrouwe.github.io/JoltPhysics/class_debug_renderer.html. From what I understand I need to subclass this class and then implement every function that is labelled virtual, is that right?

The virtual methods there are:

```
	/// Implementation of DebugRenderer interface
	virtual void						DrawLine(RVec3Arg inFrom, RVec3Arg inTo, ColorArg inColor) override;
	virtual void						DrawTriangle(RVec3Arg inV1, RVec3Arg inV2, RVec3Arg inV3, ColorArg inColor, ECastShadow inCastShadow = ECastShadow::Off) override;
	virtual Batch						CreateTriangleBatch(const Triangle *inTriangles, int inTriangleCount) override;
	virtual Batch						CreateTriangleBatch(const Vertex *inVertices, int inVertexCount, const uint32 *inIndices, int inIndexCount) override;
	virtual void						DrawGeometry(RMat44Arg inModelMatrix, const AABox &inWorldSpaceBounds, float inLODScaleSq, ColorArg inModelColor, const GeometryRef &inGeometry, ECullMode inCullMode, ECastShadow inCastShadow, EDrawMode inDrawMode) override;
	virtual void						DrawText3D(RVec3Arg inPosition, const string_view &inString, ColorArg inColor, float inHeight) override;
```

I'm familiar with how I might implement `DrawLine` and `DrawTriangle`, but why do we have have to implement both? Does this imply that there is a way to toggle a wireframe mode and a triangle mode with the debug renderer?

I'm hoping that someone can explain if I have to implement all of the virtual methods or if I can just get away with implementing `DrawLine`, also what the purpose of the Batch functions are. Thanks a lot

**Comment by jrouwe** [comment]
> From what I understand I need to subclass this class and then implement every function that is labelled virtual, is that right?

Yes

DrawTriangle expects you to draw a non-wire frame triangle, so you cannot achieve the same thing with DrawLine (although you're free to implement it that way).

The reason that CreateTriangleBatch exists is so that you can create a vertex/index buffer for a large batch of triangles. It is possible to internally route this through DrawTriangle, but for complex Shapes this is so inefficient that the frame rate suffers. Most shapes call CreateTriangleBatch multiple times to create different LOD versions of the same shape (this happens only the first time when rendering a shape) and when rendering the shape it calls DrawGeometry with the created batches. Your implementation is then expected to select the right LOD and render it. If you don't know how/don't want to calculate the right LOD, just draw the highest LOD.

**Reply by cuppajoeman** [reply]
Ok, this is good to know. My game is super simple so far and doesn't use LOD or have any way to render 3d text. 

I did have a few small follow ups: 

* What does `DrawGeometry` do?
* Do I need to implement `DrawText3D`?
* If I don't want to implement certain methods, is it fine to declare it in the subclass but have an empty function body?
* Is there a minimal set of methods that I can implement which will get basic rendering going? (which ones can I not implement)

Thanks

**Reply by jrouwe** [reply]
DrawGeometry selects the appropriate LOD and picks the corresponding triangle batch created by CreateTriangleBatch and then renders it (in other words: It renders a vertex/index buffer).

DrawText3D doesn't need to be implemented, you can implement all functions as empty functions but obviously it won't render anything then.

I think you minimally want to implement CreateTriangleBatch, DrawGeometry and DrawLine.
